tools/check-messages-en.js

```diff
@@ -0,0 +1,133 @@
+      continue;
+    }
+  }
+}
+walk(fs.opendirSync(kSrcDirName));
+
+const parseOptions = {
+  parser: {
+    parse(src) {
+      return flowParser.parse(src, {
+        // Comments can't cause UI strings to appear in the app; ignore them.
+        all_comments: false,
+        comments: false,
+
+        // We plan to use Flow enums; the parser shouldn't crash on them.
+        enums: true,
+
+        // Set `tokens: true` just to work around a mysterious error.
+        //
+        // From the doc for this option:
+        //
+        // > include a list of all parsed tokens in a top-level tokens
+        // > property
+        //
+        // We don't actually want this list of tokens. String literals do
+        // get represented in the list, but as tokens, i.e., meaningful
+        // chunks of the literal source code. They come with surrounding
+        // quotes, escape syntax, etc:
+        //
+        //   'doesn\'t'
+        //   "doesn't"
+        //
+        // What we really want is the *value* of a string literal:
+        //
+        //   doesn't
+        //
+        // and we get that from the AST.
+        //
+        // Anyway, we set `true` for this because otherwise I've been seeing
+        // `parse` throw an error:
+        //
+        //   Error: Line 72: Invalid regular expression: missing /
+        //
+        // TODO: Debug and/or file an issue upstream.
+        tokens: true,
+      });
+    },
+  },
+};
+
+// Look at all files in possibleUIStringFilePaths, and collect all string
+// literals that might represent UI strings.
+const possibleUiStringLiterals = new Set();
+possibleUIStringFilePaths.forEach(filePath => {
+  const source = fs.readFileSync(filePath).toString();
+  const ast = parse(source, parseOptions);
+
+  visit(ast, {
+    // Find nodes with type "Literal" in the AST.
+    /* eslint-disable no-shadow */
+    visitLiteral(path) {
+      // To reduce false negatives, return false when `path` definitely
+      // doesn't represent a string literal for a UI string in the app.
+
+      const { value } = path.value;
+
+      // Non-string literals: numbers, booleans, etc.
+      if (typeof value !== 'string') {
+        return false;
+      }
+
+      // String literals like 'react' in lines like
+      //   import React from 'react';
+      if (n.ImportDeclaration.check(path.parent.value)) {
+        return false;
+      }
+
+      possibleUiStringLiterals.add(value);
+
+      return this.traverse(path);
+    },
+  });
+});
+
+// Check each key ("message ID" in formatjs's lingo) against
+// possibleUiStringLiterals, and make a list of any that aren't found.
+const danglingMessageIds = Object.keys(messages_en).filter(
+  messageId => !possibleUiStringLiterals.has(messageId),
+);
+
+if (danglingMessageIds.length > 0) {
+  console.warn(
+    "Found message IDs in static/translations/messages_en.json that don't seem to be used in the app:",
+  );
+  console.warn(danglingMessageIds);
```

Comment:
For a failure result, in addition to printing output also say `process.exit(1);`. That way the script exits with status 1 (instead of 0, which is what you get if it reaches the end of the script code.)

Exit status 0 means success and anything else means failure, by convention and in particular as interpreted by the shell:
https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
so that's what we'll want for running this from `tools/test`.